### PR TITLE
Stop hanging end-to-end tests when navigating between different pages

### PR DESCRIPTION
**Pull request by mehansen**
# FRONTEND PULL REQUEST

## Related Issue

End-to-end tests have been hanging, mostly during `08-save_and_start_test_spec.js` but sometimes during `02-add_patient_spec.js`. Cypress clears local storage between test blocks, so when running with Okta, we save local storage (which includes the Okta access token) after each test block and then restore it again before each test block. I think what happens is if you are loading a new page of the app while switching to a new test block, sometimes Cypress will try to load the page while you do not have the access token restored yet. Then you get redirected to the Okta login page, which breaks the expected user flow and the test sits and waits for the expected page to load.

## Changes Proposed

* move test steps around so we're not loading a new page while local storage is being cleared and reloaded
* updated README to hopefully clarify how to run e2e locally

## Additional Information

- `02-add_patient_spec.js` used to hang sometimes, but it was before I updated our workflow to save the Cypress screenshots on timeout, so I can't tell where I need to fix things in that file. Will keep an eye out for additional timeout failures in that spec.

## Testing

Run, rerun, rerun again the end-to-end tests and verify they pass.
 
## Screenshots / Demos

n/a

### Design
- [n/a] Any UI/UX changes have a designer as a reviewer, and changes have been approved
- [n/a] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke-tested by both the engineering and design teams

### Content
- [n/a] Any content changes have been approved by content team

### Support
- [n/a] Any changes that might generate new support requests have been flagged to the support team
- [n/a] Any changes to support infrastructure have been demo'd to support team

### Security
- [n/a] Changes with security implications have been approved by a security engineer (changes to  authentication, encryption, handling of PII, etc.)
- [n/a] Any dependencies introduced have been vetted and discussed

### Documentation
- [n/a] Any changes to the startup configuration have been documented in the README

### CODE REVIEW

**Review by nathancrtr — DISMISSED** [review]
LGTM with the relatively limited testing we're able to do here - let's :shipit: and let it run a gazillion more times and see how it goes!

Also your README updates will save people from heartburn in the future so thank you for your changes there

**Review by nathancrtr — DISMISSED** [review]

**Review by emmastephenson — DISMISSED** [review]
Looks good, awesome detective work Merethe!

File: README.md [hunk 1]
```diff
@@ -148,37 +148,58 @@ E2E/Integration tests are available using [Cypress](https://www.cypress.io/).
...
+# frontend/cypresss.env.json
+{
+  "OKTA_USERNAME": "",
+  "OKTA_PASSWORD": "",
+  "OKTA_SECRET": ""
+}
 ```
 
+You'll also need address validation through Smarty to be working:
+- `REACT_APP_SMARTY_STREETS_KEY` must be set in `frontend/.env.local`
```
Comment (emyl3) [inline README.md:173]: I actually don't have this in my .env.local and it works 🤔

File: README.md [hunk 2]
```diff
@@ -148,37 +148,58 @@ E2E/Integration tests are available using [Cypress](https://www.cypress.io/).
...
+127.0.0.1 localhost.simplereport.gov
 ```
 
-The `frontend/.env.local` file has a template at `frontend/cypress/.env.e2e.sample`. Reach out to another developer to get proper values for these secrets.
-
 ```
-# /etc/hosts
-# add the following line
-127.0.0.1 localhost.simplereport.gov
+# frontend/cypresss.env.json
```
Comment (emyl3) [inline README.md:164]: nit: the file name looks like it has an extra "s" in cypress

Comment (emyl3) [reply README.md:164]: Also, just a heads up whenever we upgrade to v10.0.0 of cypress looks like they've gotten rid of support for the `.json` file for [cypress configurations](https://docs.cypress.io/guides/guides/environment-variables#Setting)
Screen Shot 2022-07-11 at 2 05 12 PM

File: README.md [hunk 3]
```diff
@@ -192,7 +213,17 @@ Now that you have those files set up, you are ready for a test run! There are a
   - This will run cypress with default values and display Cypress, API, DB, Frontend, and Nginx logs.
 -->
 
-Note that our Cypress tests expect there to only be 1 facility and will fail if there are multiple. Additionally, tests are expected to be run in order, so some spec files rely on data generated in previous specs.
+#### Troubleshooting
```
Comment (emyl3) [inline README.md:216]: ooo this is a great explanation!! thank you 😹

**Review by emyl3 — COMMENTED** [review]
Thank you for working on this, Merethe!! A couple questions/comments for you!

File: README.md (hunk 1 above)
Comment (mehansen) [reply README.md:173]: thanks for checking! I was about to ping you about this haha
Smarty is only really used in the org sign up spec and I know you need the backend credentials for that. let me try to get it to pass without the frontend key

Comment (mehansen) [reply README.md:173]: looks like it's not needed. without it, suggested addresses don't populate and our zip/state validation won't fire, but neither of these are actually needed for our e2e tests currently

File: README.md (hunk 2 above)
Comment (mehansen) [reply README.md:164]: that's correct, but the config file is different from [the env file](https://docs.cypress.io/guides/guides/environment-variables#Option-2-cypress-env-json), so I think we should be good

you can set env variables in the config file, but common practice seems to be for the config file to be for universal config things that would be included in the repo and for env to hold passwords and variables specific to local development

Comment (emyl3) [reply README.md:164]: oooo I see! thank you for the explanation 😸

File: README.md [hunk 4]
```diff
@@ -148,37 +148,58 @@ E2E/Integration tests are available using [Cypress](https://www.cypress.io/).
...
-Now that you have those files set up, you are ready for a test run! There are a few ways to run the tests from the `frontend` directory, including:
+Now that you have your environment set up, you are ready for a test run! There are a few ways to run the tests from the `frontend` directory, including:
 
 - `yarn e2e:local`
   - This will open an interactive test runner that lets you select browsers and run your desired specs.
-  - Prerequisite(s): the app must be running locally with the frontend base URL set to `http://localhost:3000` and the _dev_ profile set for the backend
+  - Prerequisite(s):
+    - the app must be running locally with the frontend base URL set to `http://localhost:3000` and backend running with the _dev_ profile
+    - `REACT_APP_OKTA_ENABLED=false` in `frontend/.env.local`
 - `yarn e2e:local:okta`
```
[the remark this answers is not in the document]
Comment (emyl3) [reply README.md:186]: it works... haha I wasn't running the frontend using the "Frontend with Okta" config in IntelliJ 😮‍💨 Thanks for your help and patience on this one Merethe!

**Review by emyl3 — APPROVED** [review]
Looks good to me! Thank you Merethe!! 😹 🎉

**Review by nathancrtr — APPROVED** [review]
Re-approving after additional commits

**Review by emmastephenson — APPROVED** [review]

**Pull request merged**